### PR TITLE
ENH: stats.trap - adding trapezoidal distribution closes #6028

### LINKED ISSUE #6028: stats.distributions does not have trapezoidal distribution
The continuous distributions in `scipy.stats` do not contain the trapezoidal distribution. We have two limiting cases, `uniform` and `triang`, but not the parent.

### PR DESCRIPTION
**Pull request by andyfaff**
Adds a trapezoidal distribution to `scipy.stats`. For those au-fait with the stats module please inform me of the relevant tests I need to add. The testing framework seems to be automatic, but the `test_stats.py` is mammoth, so I don't know where to start.
Possible work to be done (to satisfaction of `stats` guru):
- figure out the relevant statistics for the `_stats` method.
- check that the trapezoidal distribution reduces to the uniform and triangular distributions.
- add other tests
- add further documentation.

### CODE REVIEW

**Comment by andyfaff** [comment]
I noticed that `_distr_params.py` has all the continuous distributions listed. I can add `trap` there, but it's not clear what one is supposed to add. The file states "Sane parameters for stats.distributions.", but that comment is fairly cryptic.

**Comment by ev-br** [comment]
Parameters: adding an entry to _distr_params.py   runs this battery of tests: https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_continuous_basic.py#L72
Re *what* to add: `['trap, (1, 2)]` would mean `c=1, d=2`. 
These values will also be used for constructing the docstring example (this is what fails on Travis now).
(If you have an idea of how to document it better, please go ahead)

Additional tests (+1 for limiting cases) go to test_distributions.py, which has roughly a test case class per distribution.

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -4225,6 +4225,61 @@ def _stats(self, a, moments='mvsk'):
...
+    The standard form is in the range [0, 1] with c the mode.
+    The location parameter shifts the start to `loc`.
+    The scale parameter changes the width from 1 to `scale`.
+
+    %(example)s
+
+    """
+    def _rvs(self, c, d):
+        rands = self._random_state.rand(self._size)
+        return self._ppf(rands, c, d)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: This is inherited from superclass,  not needed here https://github.com/scipy/scipy/blob/master/scipy/stats/_distn_infrastructure.py#L801

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -4225,6 +4225,61 @@ def _stats(self, a, moments='mvsk'):
...
+
+    def _argcheck(self, c, d):
+        return (c >= 0) & (c <= 1) & (d >= 0) & (d <= 1) & (d >= c)
+
+    def _pdf(self, x, c, d):
+        u = 2 / (d - c + 1)
+
+        condlist = [x==1, x < c, x <= d, x > d]
+        choicelist = [0, u * x / c, u, u * (1 - x) / (1 - d)]
+        return np.select(condlist, choicelist)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Edge cases: c=0 and d=1 is allowed by argcheck above.

Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Also, why x==1 needs special casing?

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -4225,6 +4225,61 @@ def _stats(self, a, moments='mvsk'):
...
+
+    def _ppf(self, q, c, d):
+        qc, qd = self._cdf(c, c, d), self._cdf(d, c, d)
+        condlist = [q < qc, q <= qd, q > qd]
+        choicelist = [np.sqrt(q * c * (1 + d - c)),
+                      0.5 * q * (1 + d - c) + 0.5 * c,
+                      1 - sqrt((1 - q) * (d - c + 1) * (1 - d))]
+        return np.select(condlist, choicelist)
+trap = trap_gen(a=0.0, b=1.0, name="trap")
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: A voice from the bikeshed: trap or trapz?

**Comment by ev-br** [comment]
> figure out the relevant statistics for the _stats method.

You can just define `_munp(self, n, c, d)` to compute n-th central moment. Or leave it be --- this would be nice to have, but not absolutely necessary IMO.

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (andyfaff) [inline scipy/stats/_continuous_distns.py]: I need to think a little more here. If d==1 then there is a divide by zero.

File: scipy/stats/_continuous_distns.py [hunk 4]
```diff
@@ -4225,6 +4225,61 @@ def _stats(self, a, moments='mvsk'):
...
+        choicelist = [0, u * x / c, u, u * (1 - x) / (1 - d)]
+        return np.select(condlist, choicelist)
+
+    def _cdf(self, x, c, d):
+        condlist = [x==1, x < c, x <= d, x > d]
+        choicelist = [1,
+                      x**2 / c / (d - c + 1),
+                      (c + 2 * (x - c)) / (d - c + 1),
+                      1 - ((1 - x)**2 / (d - c + 1) / (1 - d))]
+        return np.select(condlist, choicelist)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: np.select is cute --- does it handle x, c or d being arrays?

Comment (andyfaff) [inline scipy/stats/_continuous_distns.py]: Best solution found so far. It handles x being array or scalar, but c and d have to be scalar.
I had a few mental gymnastics figuring out how to do this without examining if x was array or scalar.

File: scipy/stats/_continuous_distns.py (hunk 3 above)
Comment (andyfaff) [inline scipy/stats/_continuous_distns.py]: Perhaps trapz because of cumtrapz?

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Ok, I see that `triang` has the same edge case: c=0 is explicitly allowed despite a division by zero.

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: One simple option is of course just two calls to np.where.

**Comment by andyfaff** [comment]
@ev-br I added sane values to the trap entry in `_distr_params.py` (so the doctests would pass) but now I'm getting a weird error, see below. Why on earth is it generating a shape tuple of  `(1.8589710689636982, 1.461789732784883)` instead of (0.2, 0.8) like I specified in `_distr_params.py`?

> ======================================================================
ERROR: test_distributions.test_all_distributions('trap', (1.8589710689636982, 1.461789732784883), 0.01)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/anz/miniconda3/envs/dev3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/anz/Documents/Andy/programming/scipy/build/testenv/lib/python3.4/site-packages/scipy/stats/tests/test_distributions.py", line 61, in check_distribution
    D, pval = stats.kstest(dist, '', args=args, N=1000)
  File "/Users/anz/Documents/Andy/programming/scipy/build/testenv/lib/python3.4/site-packages/scipy/stats/stats.py", line 3904, in kstest
    vals = np.sort(rvs(*args, **kwds))
  File "/Users/anz/Documents/Andy/programming/scipy/build/testenv/lib/python3.4/site-packages/scipy/stats/_distn_infrastructure.py", line 855, in rvs
    raise ValueError("Domain error in arguments.")
ValueError: Domain error in arguments.

**Comment by ev-br** [comment]
(As well hidden in the traceback), this is a different set of tests, which generates random sets of parameters:
https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_distributions.py#L70

**Comment by ev-br** [comment]
You seem to be hitting quite a few ugly corners of the distributions framework...

The last remaining failure can be taken care of by just listing `trap` as a failing fit, see the list in stats/tests/test_fit.py.

I've played with it a bit to check how it fares for array-valued shapes, and it seems OK -- there's an additional simple test in https://github.com/ev-br/scipy/tree/pr/6030, can you take it over?
In  that branch, I also removed extraneous cases from the `select`: the out-of-bounds values are handled automatically by the framework, and `_pdf` does not even see `x < 0`.

From my POV, this is ready modulo blacklisting the test and renaming it to trapz.

**Comment by ev-br** [comment]
Ok, I' m going to merge it. Edge cases, c=0 and d=1 might warrant additional care, but that can be taken care of in a sequel when this sees some usage. Thanks @andyfaff.

**Pull request merged**